blackboxopt/optimization_loops/file_based_distributed.py

```diff
@@ -0,0 +1,160 @@
+
+        current_proposals = glob.glob(str(target_directory / "eval_spec_*.json"))
+        while len(current_proposals) < proposal_queue_size:
+            eval_spec_id = str(uuid4())
+            try:
+                eval_spec = optimizer.generate_evaluation_specification()
+                eval_spec.optimizer_info["eval_spec_id"] = eval_spec_id
+                with open(
+                    target_directory / f"eval_spec_{eval_spec_id}.json",
+                    "w",
+                    encoding="utf-8",
+                ) as fh:
+                    json.dump(eval_spec.to_dict(), fh)
+                current_proposals = glob.glob(
+                    str(target_directory / "eval_spec_*.json")
+                )
+            except OptimizerNotReady:
+                continue
+
+        time.sleep(0.5)
+
+
+def evaluate_specifications(
+    target_directory: PathLike,
+    evaluation_function: Callable[[EvaluationSpecification], Evaluation],
+    objectives: List[Objective],
+    timeout_s: float = float("inf"),
+    max_evaluations: Optional[int] = None,
+    catch_exceptions_from_evaluation_function: bool = False,
+    logger: Optional[logging.Logger] = None,
+):
+    """Evaluate specifications from the target directory until the `timeout_s` or
+    `max_evaluations` is reached.
+
+    Args:
+        target_directory: The directory where the evaluation specifications and results
+            are stored.
+        evaluation_function: The function that evaluates the evaluation specification.
+        objectives: The objectives reported in the evaluation function. This is used to
+            specify None values for objectives in case of an error.
+        timeout_s: The maximum time in seconds to run the optimization loop.
+        max_evaluations: The maximum number of evaluations to perform on this agent.
+        catch_exceptions_from_evaluation_function: Whether to exit on an unhandled
+            exception raised by the evaluation function or instead store their stack
+            trace in the evaluation's `stacktrace` attribute. Set to True if there are
+            spurious errors due to e.g. numerical instability that should not halt the
+            optimization loop. For more details, see the wrapper that is used internally
+            `blackboxopt.optimization_loops.utils.evaluation_function_wrapper`
+        logger: The logger to use for logging. If `None`, the default logger is used.
+    """
+    if logger is None:
+        logger = default_logger
+
+    target_directory = Path(target_directory)
+
+    _max_evaluations = init_max_evaluations_with_limit_logging(
+        max_evaluations=max_evaluations, timeout_s=timeout_s, logger=logger
+    )
+
+    start = time.time()
+    num_evaluations = 0
+    while time.time() - start < timeout_s and num_evaluations < _max_evaluations:
+        current_proposals = glob.glob(str(target_directory / "eval_spec_*.json"))
+        for eval_spec_path in current_proposals:
```

Comment:
Looping over this list runs into the risk that a worker might try to load a file that existed before the previous evaluation started, but was already processed in the mean time by a different worker. In that case, the `open` will fail in the next line.
Not that there is a 100% safe way of doing it, but I would propose to only process a single proposal from the list (e.g. pick a random one) and update it every time before a file is loaded. If this needs to be more robust, an additional `try` `except` block could catch the (hopefully) rare cases where the file is deleted between the `glob` and the `open`.